services/src/main/java/org/keycloak/protocol/oidc/endpoints/LoginStatusIframeEndpoint.java

```diff
@@ -60,12 +63,21 @@ public Response preCheck(@QueryParam("client_id") String clientId, @QueryParam("
             ClientModel client = session.clients().getClientByClientId(realm, clientId);
             if (client != null && client.isEnabled()) {
                 Set<String> validWebOrigins = WebOriginsUtils.resolveValidWebOrigins(session, client);
-                validWebOrigins.add(UriUtils.getOrigin(uriInfo.getRequestUri()));
+                String requestOrigin = UriUtils.getOrigin(uriInfo.getRequestUri());
+                validWebOrigins.add(requestOrigin);
                 if (validWebOrigins.contains("*") || validWebOrigins.contains(origin)) {
                     return Response.noContent().build();
                 }
+                if (logger.isDebugEnabled()) {
+                    logger.debugf("client %s does not allow origin=%s for requestOrigin=%s (as determined by hostname settings), init will return a 403", clientId, origin, requestOrigin);
+                }
+            } else {
+                if (logger.isDebugEnabled()) {
+                    logger.debugf("client %s does not exist or not enabled, init will return a 403", clientId);
+                }
```

Comment:
Unrelated to this implementation, but does `debugf()` itself not check if debugging is enabled? Also, why is this not needed for the `logger.debug()` call below?

Reply:
> Unrelated to this implementation, but does debugf() itself not check if debugging is enabled?

It does, and it is checked prior to the formatting of the string, so it is unnecessary here.

> Also, why is this not needed for the logger.debug() call below?

Because it is an exceptional case, so I don't think there's any concern about it being on a critical path for performance.

Updated so that none of them use an explicit check.